devdoc/cancellation_token_requirements.md

```diff
@@ -0,0 +1,138 @@
+
+**SRS_CANCELLATION_TOKEN_04_008: [** `cancellation_token_is_canceled` shall assign `true` to `*canceled` if the token has been canceled and `false` otherwise. **]**
+
+**SRS_CANCELLATION_TOKEN_04_009: [** `cancellation_token_is_canceled` shall return `0` when successful. **]**
+
+## cancellation_token_register_notify
+
+```c
+MOCKABLE_FUNCTION(, THANDLE(CANCELLATION_TOKEN_REGISTRATION), cancellation_token_register_notify, THANDLE(CANCELLATION_TOKEN), cancellation_token, TCALL_DISPATCHER_TARGET_FUNC_TYPE_NAME(CANCELLATION_TOKEN_CANCEL_CALL), on_cancel, void*, context);
+```
+
+`cancellation_token_register_notify` registers `on_cancel` to be called when the token is canceled.
+
+**SRS_CANCELLATION_TOKEN_04_010: [** `cancellation_token_register_notify` shall fail and return `NULL` when `cancellation_token` is `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_011: [** `cancellation_token_register_notify` shall fail and return `NULL` when `on_cancel` is `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_012: [** `cancellation_token_register_notify` shall fail and return `NULL` when any underlying call fails. **]**
+
+**SRS_CANCELLATION_TOKEN_04_013: [** `cancellation_token_register_notify` shall call `TCALL_DISPATCHER_REGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL)` to register `on_cancel` with the dispatcher. **]**
+
+**SRS_CANCELLATION_TOKEN_04_014: [** `cancellation_token_register_notify` shall call `THANDLE_MALLOC(CANCELLATION_TOKEN_REGISTRATION)` to allocate a `CANCELLATION_TOKEN_REGISTRATION` handle. **]**
+
+**SRS_CANCELLATION_TOKEN_04_015: [** `cancellation_token_register_notify` shall initialize and return a valid `THANDLE(CANCELLATION_TOKEN_REGISTRATION)` when successful. **]**
+
+**SRS_CANCELLATION_TOKEN_04_016: [** `cancellation_token_register_notify` shall call `TCALL_DISPATCHER_UNREGISTER_TARGET(CANCELLATION_TOKEN_CANCEL_CALL)` when an error occurs to undo the callback registration. **]**
+
+## cancellation_token_cancel
+
+```c
+MOCKABLE_FUNCTION(, int, cancellation_token_cancel, THANDLE(CANCELLATION_TOKEN), cancellation_token);
+```
+
+`cancellation_token_cancel` sets the state of the cancellation token to be "canceled" and invokes all the callbacks that have been registered via `cancellation_token_register_notify`.
+
+**SRS_CANCELLATION_TOKEN_04_017: [** `cancellation_token_cancel` shall fail and return a non-zero value if `cancellation_token` is `NULL`. **]**
+
+**SRS_CANCELLATION_TOKEN_04_018: [** `cancellation_token_cancel` shall fail and return a non-zero value if the state of the token is already "canceled". **]**
+
+**SRS_CANCELLATION_TOKEN_04_019: [** `cancellation_token_cancel` shall set the state of the token to be "canceled". **]**
+
+**SRS_CANCELLATION_TOKEN_04_020: [** `cancellation_token_cancel` shall return `0` when successful. **]**
+
+**SRS_CANCELLATION_TOKEN_04_021: [** `cancellation_token_cancel` shall invoke all callbacks that have been registered on the token via `cancellation_token_register_notify`. **]**
```

Comment:
>invoke all callbacks [](http://example.com/codeflow?start=71&length=20)

It does that through TCALL_DISPATCHER means, right? Would be nice to have that explicitly written here.